test/strings/types.jl

```diff
@@ -2,332 +2,347 @@
-            @test prevind(ss, i+1, j) == prevind(s, i+1, j)
-            @test nextind(ss, i, j) == nextind(s, i, j)
+
+    @testset "issue #5870" begin
+        @test !occursin(Regex("aa"), SubString("",1,0))
+        @test occursin(Regex(""), SubString("",1,0))
+    end
+    @testset" isvalid, length, prevind, nextind for SubString{String}" begin
+        s = "lorem ipsum"
+        sdict = Dict(
+            SubString(s, 1, 11)  => "lorem ipsum",
+            SubString(s, 1, 6)   => "lorem ",
+            SubString(s, 1, 0)   => "",
+            SubString(s, 2, 4)   => "ore",
+            SubString(s, 2, 11)  => "orem ipsum",
+            SubString(s, 15, 14) => "",
+        )
+        for (ss, s) in sdict
+            @test ncodeunits(ss) == ncodeunits(s)
+            for i in -2:13
+                @test isvalid(ss, i) == isvalid(s, i)
+            end
+            for i in 1:ncodeunits(ss), j = i-1:ncodeunits(ss)
+                @test length(ss, i, j) == length(s, i, j)
+            end
+        end
+        for (ss, s) in sdict
+            @test length(ss) == length(s)
+            for i in 0:ncodeunits(ss), j = 0:length(ss)+1
+                @test prevind(ss, i+1, j) == prevind(s, i+1, j)
+                @test nextind(ss, i, j) == nextind(s, i, j)
+            end
+            @test_throws BoundsError prevind(s, 0)
+            @test_throws BoundsError prevind(ss, 0)
+            @test_throws BoundsError nextind(s, ncodeunits(ss)+1)
+            @test_throws BoundsError nextind(ss, ncodeunits(ss)+1)
         end
-        @test_throws BoundsError prevind(s, 0)
-        @test_throws BoundsError prevind(ss, 0)
-        @test_throws BoundsError nextind(s, ncodeunits(ss)+1)
-        @test_throws BoundsError nextind(ss, ncodeunits(ss)+1)
     end
-end
 
-# proper nextind/prevind/thisind for SubString{String}
-let rng = MersenneTwister(1), strs = ["∀∃∀"*String(rand(rng, UInt8, 40))*"∀∃∀",
-                                      String(rand(rng, UInt8, 50))]
-    for s in strs
-        a = 0
-        while a <= ncodeunits(s)
-            a = nextind(s, a)
-            b = a - 1
-            while b <= ncodeunits(s)
-                ss = SubString(s, a:b)
-                s2 = s[a:b]
-                @test ncodeunits(ss) == ncodeunits(s2)
-                for i in 0:ncodeunits(ss)+1
-                    @test thisind(ss, i) == thisind(s2, i)
-                end
-                for i in 0:ncodeunits(ss)
-                    @test nextind(ss, i) == nextind(s2, i)
-                    for j in 0:ncodeunits(ss)+5
-                        if j > 0 || isvalid(ss, i)
-                            @test nextind(ss, i, j) == nextind(s2, i, j)
+    @testset "proper nextind/prevind/thisind for SubString{String}" begin
+        rng = MersenneTwister(1)
+        strs = ["∀∃∀"*String(rand(rng, UInt8, 40))*"∀∃∀",
+                                          String(rand(rng, UInt8, 50))]
+        for s in strs
```

Comment:
These tests look really hard to debug when failing. Might be good to generate `strs` outside the testset and change it to something like:
```suggestion
    rng = MersenneTwister(1)
    strs = ["∀∃∀"*String(rand(rng, UInt8, 40))*"∀∃∀",
                                      String(rand(rng, UInt8, 50))]
    @testset "proper nextind/prevind/thisind for SubString{String}: $(repr(s))" for s in strs
```